test/test_unified_format.py

```diff
@@ -0,0 +1,34 @@
+# Copyright 2020-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+
+sys.path[0:0] = [""]
+
+from test import unittest
+from test.unified_format import generate_test_classes
+
+
+_TEST_PATH = os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), 'unified-test-format')
+
+globals().update(generate_test_classes(
+    os.path.join(_TEST_PATH, 'valid-pass'),
+    module=__name__,
+    class_name_prefix='UnifiedTestFormat'))
```

Comment:
This creates test classes bearing the name `test.test_unified_format.TestUnifiedFormat*`

Reply:
How many test classes does this generate? Does a class refer to a single json file or a single directory?

Reply:
The current set up creates one class per JSON file.

Reply:
Thoughts on this? Do you think this is too many classes? I find the test names are more sensible when we do this (ignore ERRORs):
```
test_Executing_a_watch_helper_on_a_MongoClient_results_in_notifications_for_changes_to_all_collections_in_all_databases_in_the_cluster_ (test.test_unified_format.TestUnifiedTestFormatValidPassPocChangeStreams) ... ERROR
test_Test_consecutive_resume (test.test_unified_format.TestUnifiedTestFormatValidPassPocChangeStreams) ... ERROR
test_A_failed_find_event (test.test_unified_format.TestUnifiedTestFormatValidPassPocCommandMonitoring) ... ERROR
test_A_successful_find_event_with_a_getmore_and_the_server_kills_the_cursor (test.test_unified_format.TestUnifiedTestFormatValidPassPocCommandMonitoring) ... ERROR
test_Aggregate_with_$listLocalSessions (test.test_unified_format.TestUnifiedTestFormatValidPassPocCrud) ... ERROR
test_BulkWrite_with_mixed_ordered_operations (test.test_unified_format.TestUnifiedTestFormatValidPassPocCrud) ... ERROR
test_InsertMany_continue-on-error_behavior_with_unordered_(duplicate_key_in_requests) (test.test_unified_format.TestUnifiedTestFormatValidPassPocCrud) ... ERROR
test_ReplaceOne_prohibits_atomic_modifiers (test.test_unified_format.TestUnifiedTestFormatValidPassPocCrud) ... ERROR
test_readConcern_majority_with_out_stage (test.test_unified_format.TestUnifiedTestFormatValidPassPocCrud) ... ERROR
test_Delete_when_length_is_10 (test.test_unified_format.TestUnifiedTestFormatValidPassPocGridfs) ... ERROR
test_Download_when_an_intermediate_chunk_is_missing (test.test_unified_format.TestUnifiedTestFormatValidPassPocGridfs) ... ERROR
test_Download_when_files_entry_does_not_exist (test.test_unified_format.TestUnifiedTestFormatValidPassPocGridfs) ... ERROR
test_Download_when_there_are_three_chunks (test.test_unified_format.TestUnifiedTestFormatValidPassPocGridfs) ... ERROR
test_Upload_when_length_is_5 (test.test_unified_format.TestUnifiedTestFormatValidPassPocGridfs) ... ERROR
test_Aggregate_succeeds_after_InterruptedAtShutdown (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableReads) ... ERROR
test_Find_fails_on_first_attempt (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableReads) ... ERROR
test_Find_fails_on_second_attempt (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableReads) ... ERROR
test_Find_succeeds_on_second_attempt (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableReads) ... ERROR
test_ListDatabases_succeeds_on_second_attempt (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableReads) ... ERROR
test_FindOneAndUpdate_is_committed_on_first_attempt (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableWrites) ... ERROR
test_FindOneAndUpdate_is_never_committed (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableWrites) ... ERROR
test_FindOneAndUpdate_is_not_committed_on_first_attempt (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableWrites) ... ERROR
test_InsertMany_succeeds_after_PrimarySteppedDown (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableWrites) ... ERROR
test_InsertOne_fails_after_connection_failure_when_retryWrites_option_is_false (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableWrites) ... ERROR
test_InsertOne_fails_after_multiple_retryable_writeConcernErrors (test.test_unified_format.TestUnifiedTestFormatValidPassPocRetryableWrites) ... ERROR
test_Dirty_explicit_session_is_discarded (test.test_unified_format.TestUnifiedTestFormatValidPassPocSessions) ... ERROR
test_Server_supports_explicit_sessions (test.test_unified_format.TestUnifiedTestFormatValidPassPocSessions) ... ERROR
test_Server_supports_implicit_sessions (test.test_unified_format.TestUnifiedTestFormatValidPassPocSessions) ... ERROR
test_Client_side_error_in_command_starting_transaction (test.test_unified_format.TestUnifiedTestFormatValidPassPocTransactions) ... ERROR
test_create_index_on_a_non-existing_collection (test.test_unified_format.TestUnifiedTestFormatValidPassPocTransactions) ... ERROR
test_explicitly_create_collection_using_create_command (test.test_unified_format.TestUnifiedTestFormatValidPassPocTransactions) ... ERROR
test_withTransaction_and_no_transaction_options_set (test.test_unified_format.TestUnifiedTestFormatValidPassPocTransactionsConvenientApi) ... ERROR
test_withTransaction_explicit_transaction_options (test.test_unified_format.TestUnifiedTestFormatValidPassPocTransactionsConvenientApi) ... ERROR
test_withTransaction_inherits_transaction_options_from_client (test.test_unified_format.TestUnifiedTestFormatValidPassPocTransactionsConvenientApi) ... ERROR
test_withTransaction_inherits_transaction_options_from_defaultTransactionOptions (test.test_unified_format.TestUnifiedTestFormatValidPassPocTransactionsConvenientApi) ... ERROR
```

Reply:
Right now many of our test files have multiple test classes in addition to the spec tests. Won't this make it impossible to run a whole test suite without running the entire module? For example, in test_encryption we have:
```
git grep '^class ' -- test/test_encryption.py | tee  
test/test_encryption.py:70:class TestAutoEncryptionOpts(PyMongoTestCase):
test/test_encryption.py:114:class TestClientOptions(PyMongoTestCase):
test/test_encryption.py:132:class EncryptionIntegrationTest(IntegrationTest):
test/test_encryption.py:176:class TestClientSimple(EncryptionIntegrationTest):
test/test_encryption.py:268:class TestClientMaxWireVersion(IntegrationTest):
test/test_encryption.py:310:class TestExplicitSimple(EncryptionIntegrationTest):
test/test_encryption.py:461:class TestSpec(SpecRunner):
test/test_encryption.py:599:class TestDataKeyDoubleEncryption(EncryptionIntegrationTest):
test/test_encryption.py:725:class TestExternalKeyVault(EncryptionIntegrationTest):
test/test_encryption.py:791:class TestViews(EncryptionIntegrationTest):
test/test_encryption.py:812:class TestCorpus(EncryptionIntegrationTest):
test/test_encryption.py:967:class TestBsonSizeBatches(EncryptionIntegrationTest):
test/test_encryption.py:1068:class TestCustomEndpoint(EncryptionIntegrationTest):
test/test_encryption.py:1230:class AzureGCPEncryptionTestMixin(object):
test/test_encryption.py:1294:class TestAzureEncryption(AzureGCPEncryptionTestMixin,
test/test_encryption.py:1320:class TestGCPEncryption(AzureGCPEncryptionTestMixin,
```

So we can run only the spec tests like this:
```
python setup.py test -s test.test_encryption.TestSpec
```

Reply:
Good point. You are correct, doing this is not possible with the current implementation.